website/source/guides/security/vault-pki-integration.html.md

```diff
@@ -0,0 +1,614 @@
+  destination = "/opt/nomad/agent-certs/agent.crt"
+
+  # This is the permission to render the file. If this option is left
+  # unspecified, Consul Template will attempt to match the permissions of the
+  # file that already exists at the destination path. If no file exists at that
+  # path, the permissions are 0644.
+  perms       = 0700
+
+  # This is the optional command to run when the template is rendered. The
+  # command will only run if the resulting template changes. 
+  command     = "systemctl reload nomad"
+}
+
+template {
+  source      = "/opt/nomad/templates/agent.key.tpl"
+  destination = "/opt/nomad/agent-certs/agent.key"
+  perms       = 0700
+  command     = "systemctl reload nomad"
+}
+
+template {
+  source      = "/opt/nomad/templates/ca.crt.tpl"
+  destination = "/opt/nomad/agent-certs/ca.crt"
+  command     = "systemctl reload nomad"
+}
+
+# The following template stanzas are for the CLI certs
+
+template {
+  source      = "/opt/nomad/templates/cli.crt.tpl"
+  destination = "/opt/nomad/cli-certs/cli.crt"
+}
+
+template {
+  source      = "/opt/nomad/templates/cli.key.tpl"
+  destination = "/opt/nomad/cli-certs/cli.key"
+}
+```
+
+!> Note: we have hard-coded the token we created into the Consul Template
+configuration file. Although we can avoid this by assigning it to the
+environment variable `VAULT_TOKEN`, this method can still pose a security
+concern. The recommended approach is to securely introduce this token to Consul
+Template. To learn how to accomplish this, see [Secure
+Introduction][secure-introduction].
+
+* Please also note we have applied file permissions `0700` to the `agent.crt`
+  and `agent.key` since only the root user should be able to read those files.
+  Any other user using the Nomad CLI will be able to read the CLI certs and key
+  that we have created for them along with intermediate CA cert.
+
+
+### Step 12: Start the Consul Template Service
+
+Start the Consul Template service on each node:
+
+```shell
+$ sudo systemctl start consul-template
+```
+You can quickly confirm the appropriate certs and private keys were generated in
+the `destination` directory you specified in your Consul Template configuration
+by listing them out:
+
+```
+$ ls /opt/nomad/agent-certs/ /opt/nomad/cli-certs/
+/opt/nomad/agent-certs/:
+agent.crt  agent.key  ca.crt
+
+/opt/nomad/cli-certs/:
+cli.crt  cli.key
+```
+
+### Step 13: Configure Nomad to Use TLS
+
+Add the following [tls stanza][nomad-tls-stanza] to the configuration of all
+Nomad agents (servers and clients) in the cluster (configuration file located at
+`/etc/nomad.d/nomad.hcl` in this example):
+
+```
+tls {
+  http = true
+  rpc  = true
+
+  ca_file   = "/opt/nomad/agent-certs/ca.crt"
+  cert_file = "/opt/nomad/agent-certs/agent.crt"
+  key_file  = "/opt/nomad/agent-certs/agent.key"
+
+  verify_server_hostname = true
+  verify_https_client    = true
+}
+```
+Reload Nomad's configuration:
+
+```shell
+$ systemctl reload nomad
```

Comment:
Unfortunately there's one last bit of complexity here for upgrading _existing_ clusters to TLS: https://www.nomadproject.io/guides/security/securing-nomad.html#migrating-a-cluster-to-tls

Basically:

As soon as servers are reloaded they will start rejecting connections from clients. In non-trivial clusters if it takes a while for reloads to happen across the cluster some clients without TLS loaded may miss heartbeats when the quorum of servers have TLS loaded. The worst case scenario is that clients are unable to connect to servers for long enough they become `lost` and rescheduling occurs.

So on servers you need to do an ugly double-reload dance:

1. Set [`tls.rpc_upgrade_mode = true`](https://www.nomadproject.io/docs/configuration/tls.html#rpc_upgrade_mode) on the servers.
2. Reload all nodes.
3. Set `tls.rpc_ugprade_mode = false` (or remove the line).
4. Reload servers.

Reply:
Thank you for pointing this out!